### DISCUSSION TITLE
nrf24l01 Module not Working with Raspberry Pi Pico W

### DISCUSSION THREAD

**Discussion by Marin-Johnson** [micropython/micropython#10454] [RP2040 / Pico] [answered]
I can use the [nrf24l01 wireless module](https://www.sparkfun.com/products/705) just fine with a Raspberry Pi Pico, but when I try to run the same code on the Raspberry Pi Pico **W**, it fails to recognize the device. I am using [this][1] library, and I get this error:

    Traceback (most recent call last):
    File "<stdin>", line 31, in <module>
    File "nrf24l01.py", line 78, in __init__
    OSError: nRF24L01+ Hardware not responding

The boards are supposed to have the same footprint, so what is the difference here? I am using the following pins:

{"spi": 0, "miso": 4, "mosi": 7, "sck": 6, "csn": 14, "ce": 17}

How can I use these modules with the Pico W?

  [1]: https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/radio/nrf24l01

**Comment by peterhinch** [comment]
Please indicate what pins you are allocating to the device.

**Reply by Marin-Johnson** [reply]
I am using the following pins:

{"spi": 0, "miso": 4, "mosi": 7, "sck": 6, "csn": 14, "ce": 17}

**Reply by peterhinch** [reply]
OK, I don't think the problem is the pinout. My guess is that it's because there are two 2.4GHz radio antennas in very close proximity. The NRF's sensitive receiver is next to a relatively powerful transmitter: maybe the NRF doesn't like the WiFi shouting in its ear. This could be proved by turning off the WiFi - I'm not sure how to do this, try the Raspberry Pi docs.

Other ideas are to try a different channel for the NRF.

If I'm right about the cause, a probable solution is this. Sparkfun do a very good NRF24L01 with a reverse SMA connector. A suitable screened cable and a duck antenna would shift the NRF antenna away from the WiFi's.

**Reply by Marin-Johnson** [reply]
Thank you! Will try these suggestions.

**Comment by peterhinch** [comment]
Another thought is to turn off access point support: this should quieten the WiFi:
```python
  import network
  sta_if = network.WLAN(network.STA_IF)  # Could deactivate this too?
  ap = network.WLAN(network.AP_IF)  # create access-point interface
  ap.active(False)    # deactivate the interface
```

**Reply by Marin-Johnson** [reply]
I just tried this and keep getting the same error: `nRF24L01+ Hardware not responding`. I also tried adding a delay after turning off the access point in case it took a while to shut down. Thank you for your suggestions.

**Reply by Marin-Johnson** [reply]
@peterhinch I also tried moving the module farther from the pico, but no change.

**Comment by peterhinch** [comment, marked as answer]
I've finally found time to try this and I can't replicate the fault. I took the following steps:
 1. Set up a Pico with an NRF24L01 on standard pins, copied the driver and `nrf24l01test.py` to the card.
 2. Set up a Pico W with an NRF24L01 on the pins you are using.
 3. Copied the driver and `nrf24l01test.py` to the card.
 4. Edited `nrf24l01test.py` on the card to match your pinout.
 5. On the Pico W ran `nrf24l01test.responder`.
 6. On the Pico ran `nrf24l01test.initiator`. This ran to completion with no errors.

I did nothing with regards to the Pico W's radio. I ran both NRF24L01 boards from 3.3V.

Conclusions:
 1. Your pinout is fine.
 2. My ideas about the radio are wrong, at least with the radio in its initial state.
 3. Something is adrift with your setup. Are you sure your NRF board is still working? I have come across some very dodgy boards from eBay. I use Sparkfun boards which are excellent.

**Reply by Marin-Johnson** [reply]
Thank you for your help. I am currently using the SparkFun NRF Boards, I have previously tried some cheap ones off amazon and they were junk. I will try a new Pico W and NRF module.

**Reply by Marin-Johnson** [reply]
I just found the problem: I was initializing the SPI Interface wrong. When I copied the SPI initialization code from the sample `nrf24l01test.py`, it worked! I was previously using some code I got [here](https://coffeebreakpoint.com/micropython/how-to-connect-a-nrf24l01-transceiver-to-your-raspberry-pi-pico/). It now runs fine on the Pico and the Pico W. It seems that the old code works on the Pico but not on the Pico W:

    cfg = {"spi": 0, "miso": 4, "mosi": 7, "sck": 6, "csn": 14, "ce": 17}

    csn = Pin(cfg["csn"], mode=Pin.OUT, value=1)
    ce = Pin(cfg["ce"], mode=Pin.OUT, value=0)
    nrf = NRF24L01(SPI(cfg["spi"]), csn, ce, payload_size=32)

And the sample code from `nrf24l01test.py` works on both the Pico and the Pico W:

    spi = SPI(0, sck=Pin(6), mosi=Pin(7), miso=Pin(4))
    cfg = {"spi": spi, "miso": 4, "mosi": 7, "sck": 6, "csn": 14, "ce": 17}

    csn = Pin(cfg["csn"], mode=Pin.OUT, value=1)
    ce = Pin(cfg["ce"], mode=Pin.OUT, value=0)
    spi = cfg["spi"]
    nrf = NRF24L01(spi, csn, ce, payload_size=32)


Both work flawlessly now. Thank you very much for all your time and help.

**Reply by peterhinch** [reply]
I'm glad you got it working.

I don't know why the Pico and Pico W differ, but I have found it advisable always to specify pins for SPI, UART and I2C interfaces. There are defaults, but without delving into Raspberry Pi docs it's not obvious what they are. It's simplest to specify pins.

**Reply by zamora1221** [reply]
where can i find the code that works on the pico w? the nrf24l01test.py

**Reply by peterhinch** [reply]
See [the official library](https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/radio/nrf24l01).

**Reply by NMUKHERJEE18** [reply]
I am also having the same Issue. it appears that the nrf24l01 module only works on certain Picos. I tried it on two regular raspberry pi picos and it did not work, but when I tried it with a Raspberry pi pico H it worked. I keep getting the error : OSError: nRF24L01+ Hardware not responding on both of them.

**Reply by Marin-Johnson** [reply]
What board are you using?

**Reply by NMUKHERJEE18** [reply]
This one right here: https://www.amazon.com/gp/product/B08LSPZHT8/ref=ppx_yo_dt_b_asin_title_o00?ie=UTF8&psc=1

**Reply by Marin-Johnson** [reply]
Yeah I never got those work reliability. Try the SparkFun ones [here](https://www.amazon.com/WRL-00705-Development-SparkFun-Transceiver-Breakout/dp/B092PTXGJ9/ref=mp_s_a_1_1?crid=2L8XV2SOHD0R1&dib=eyJ2IjoiMSJ9.3Q5WqqVOmChoPyfY47GGQUaO_uu1I0TVN_8-IWFYNGuUaL46C6NjqAN_q6t7IHftR3O3EZDtxy-VqUdP_9GHM3MhCSPXD7M4eFxXisQLU1Xgoqie_cHP3X6FhPSpG02dW589bdes_nzudkA08mX9rcELULoriJ0dlDn_mGYPfbQghu-ZGF6K3saK-2GpG8bUoKFmfkHmL86ycAcFNbBafQ.Yr0TXvGSW1CqFcu-7ScFAblf8-w88wditYSZo1-vOAM&dib_tag=se&keywords=sparkfun+NRF24L01&qid=1711738068&sprefix=sparkfun+nrf24l01%2Caps%2C131&sr=8-1). But yikes they used to be a lot cheaper, until SparkFun discontinued them.

**Reply by NMUKHERJEE18** [reply]
Will check them out. Thanks

**Reply by Rehcerphs** [reply]
I ran into this issue after updating the firmware on my Pico. The NRF module was working fine before I updated Pico's firmware. I actually was trying to use between two Pico, and one was flashed with older firmware from 2021 and was working, and the second pico was fresh out the box and I had just flashed with new firmware. This second Pico wasn't working with the NRF module, so I switched it out with an ESP32.  Had working communication between the Pico and ESP32 until I flashed new firmware on the first Pico, and now I'm back where I started.
